### ISSUE TITLE
Build system improvements

### ISSUE THREAD

**Issue by spxtr** [kubernetes/kubernetes#30600]
The current build system is several thousand lines of confusing bash that breaks in subtle (#30546, for example) and unexpected ways.

I would like to move toward this world:
- If a developer is familiar with the go tool and has their environment set up, they should be able to do a local, iterative build. It should be clear what commands are actually being run.
- If necessary, a developer should be able to do a fully dockerized build. Inside the container, the dockerized build should be identical to the above build.

Right now we are in an awkward in-between, where parts of the build happen locally and parts happen in the build container. I have the start of a PR that accomplishes this over at #27746, but I think it's worth opening up this issue to discuss other ideas.

Bazel, anyone? What about deprecating the entire build folder and starting from scratch so that we're not forced to honor every environment variable that snuck in over the last few years?

cc @thockin @david-mcmahon @jbeda @ixdy

**Comment by jbeda** [comment]
@thockin and I were just talking about this on Friday.

Heh -- this is how the `build/` folder started.  It is still mostly true.    The idea was that `build/*` would set up the environment so that you could run the standard build stuff in there if you wanted to.  

Much of the complexity in `build/` is dealing with copying files in and out of the docker container appropriately to support remote (or remote-ish) Docker engines.  The core scenario here is Docker on the mac (etiher Docker for Mac or docker-machine) which support some "native" support for mounting host directories but it is really slow.

The original plan was to have `hack/` not depend on `build/` at all.  This is no longer true and creates problems. Personally I'm dismayed by things like `hack/update-generated-protobuf.sh` which plays around with hacking up the build container in strange ways and running docker directly.

Things that I think we can do to improve the situation:
* Factor out the release stuff from `build`.  We can reduce it to *just* setting up and using the docker build container.
* Ensure that nothing in `hack` depends on `build` at all.
* Reduce the number of magic scripts in `hack`. Right there are all sorts of scripts to run tests, ensure static files are built, deal with cross compiliation, etc.  We need to clean that up.

---
One thing that complicates all of this is the fact that with the generated file stuff that @thockin just put in we are now writing files all over the source tree.  This means we need to copy back more than just the stuff we've isolated into `_output`.  That is a real drag.  Choices are to (a) mount the source into the container (and give up true remote docker engine support) or (b) do some fancy rsyncing back and forth (write it ourselves explicitly or use something like [docker-sync](https://duske.me/performant-docker-container-sync-with-docker-sync/)). 

Note that file operations are *much* slower on a host mount with Docker for Mac:

```
host$ time find . > /dev/null


real	0m0.080s
user	0m0.011s
sys	0m0.047s

docker$ time find . > /dev/null

real	0m6.189s
user	0m0.090s
sys	0m0.500s
```

For pure metadata operations like this native filesystem support is ~80 times faster than using Docker for Mac osxfs.

**Comment by jbeda** [comment]
Another problem is that we have a *lot* of binaries to build.  We don't have a good mapping of when to build which binaries and for which platforms.  That should probably be part of the `hack` (and maybe release) clean up.

**Comment by jbeda** [comment]
And I'm not a fan of bazel in this situation. It is yet another prerequisite that users will have to install to get building.  Personally, it seemed like a huge distraction when I was looking to build TensorFlow.  Using Bazel also works very very strongly against your first point.  It is very different and high overhead compared to "just use the go tool".

**Comment by spxtr** [comment]
Not only is it a pain to copy to/from docker machine on mac, but it's also a bother to use volumes on linux because by default any  new changes will be owned by root (#25525). I'm not sure what the best answer is there, since I haven't done much development on my mac.

I totally agree with your opinion on bazel. It was a huge hassle to install on my non-ubuntu workstation and it made installing tensorflow frustrating.

> Factor out the release stuff from build. We can reduce it to just setting up and using the docker build container.

#24629

also cc @krousey

**Comment by david-mcmahon** [comment]
Agreed.  build and hack are in desperate need of an overhaul.  Certainly breaking up the monolith, drawing clear lines and removing the inter-dependencies as well as factoring out well defined bits is a good start.   

Getting the release out of the build is already in progress and tracked in #16529.

**Comment by MHBauer** [comment]
What's needed to move forward? Reviews?

**Comment by jbeda** [comment]
Right now I'm working on making the docker build incremental.  I'm fixing up https://github.com/kubernetes/kubernetes/pull/28518 after Tim's changes.

**Comment by MHBauer** [comment]
@jbeda have been subscribed and watching. Do you want me to wait until you are done rebasing or start digging in?

**Comment by jbeda** [comment]
You should start. I'm still testing strategies.

**Comment by luxas** [comment]
It should also be possible to choose build image per arch, ref: https://github.com/kubernetes/kubernetes/pull/30659#issuecomment-240075801

**Comment by spxtr** [comment]
What about cmake?

I'd really like to move off of bash, but mixing bash and make is even worse. It'd be great if we could find a single system that works well for us. If not, well... bash it is.

**Comment by david-mcmahon** [comment]
cmake, now that's a name I've not heard for a long time.  No comment on it as a replacement though.

Makefiles with bash is quite a common pairing, certainly in OSS.  The problem isn't inherently with bash.  bash can be written and maintained in a way that makes it suck less for people, but like anything it can get quite unwieldy too without TLC along the way.

**Comment by thockin** [comment]
> Bazel, anyone?

Last I looked it was not really viable for Go, and besides that is Java.

> What about deprecating the entire build folder and starting from scratch

I think that is the only viable path

> Docker on the mac ... is really slow.

I have only the smallest smidgeon of sympathy.  If the choice is a clean build which runs slowly on Mac or a tangle of hacks upon hacks which is fast, well, my choice is clear.  But then, I am not a Mac user.

> I'm dismayed by things like hack/update-generated-protobuf.sh which plays around with hacking up the build container in strange ways and running docker directly.

+10 - we should make it run "local" and tell people that they need to install all of our esoteric build tools, or else they should use the dockerized build.  And we should make the dockerized build the default.

> or (b) do some fancy rsyncing

Please don't make Linux builds suck just because Mac builds have to suck.

> Another problem is that we have a lot of binaries to build.

hyperkube was supposed to fix this, but there's apparently issues...

> a bother to use volumes on linux because by default any new changes

Only directories *created by docker* as part of making a volume mount end up root-owned.  This is tractable if we are thoughtful about what we mount and where.

> What about cmake?

I don't know cmake well.  I knew it a little bit once.   It seems vastly more complicated, and frankly if we are going to adopt a new tool, I'd rather do bazel.  At least I know the design of it.

> bash can be written and maintained in a way that makes it suck less for people

A non-trivial part of why our bash sucks is MacOS support.  Sorry to be a broken record.

**Comment by thockin** [comment]
It would be great to EOL the vestigial hack/* scripts, but I am worried about all the blog posts that say to run `hack/build-go.sh`

**Comment by david-mcmahon** [comment]
> `hack/build-go.sh`

Seems easy enough to have it either print out a message or just redirect to the new shiny thing.

**Comment by thockin** [comment]
That's what it does, but it is still litter.

**Comment by ixdy** [comment]
I'm planning to dedicate cycles towards improving the build situation, though I'm still getting up to speed on understanding how it all works. @jbeda @thockin @spxtr @MHBauer and other interested parties - should we do a hangout/zoom/something at some point to share knowledge and current state?

**Comment by thockin** [comment]
happy to dump what I know.

**Comment by MHBauer** [comment]
@ixdy I'll show up. Anything scheduled? Send out an email?

**Comment by jbeda** [comment]
Wrt OS X support -- my latest set of changes (#30787) have things working pretty well.  There is one issue with a Jenkins machine that I'm tracking down.  The only Mac specific hacks are around gnu/bsd/old tools.  This is similar to dealing with older versions of Docker.  Data is copied in and out of the container using rsync with minimal cost.  I honestly don't think we need to throw out most of the stuff in `build/`.  (We should factor out the release stuff but I understand that is happening.)

If we are going to go "docker only" or "docker first" we should make sure that this is well communicated.  Doing anything in Docker was super controversial when this stuff was first implemented.

Happy to sit down whenever and give a brain dump.

**Comment by spxtr** [comment]
I strongly suggest against going "docker only" as I understand the term. Developers should not be using docker for their builds unless they want to.

**Comment by thockin** [comment]
I just disagree.  Docker is how we get a hermetic, repeatable build.  That
should be the default.

**Comment by MHBauer** [comment]
Everyone should be using containers for build.

**Comment by duglin** [comment]
+1

Totally biased example but it won me over..... when I first started to use Docker I basically got a Ubuntu VM, installed Docker, git cloned the docker src and said “make binary” (basically). IMO that should be the bar for Kube developer’s UX.  As today I think using Docker for Kube’s build is really only option for meeting that goal.

-Doug

**Comment by krousey** [comment]
> I strongly suggest against going "docker only" as I understand the term. Developers should not be using docker for their builds unless they want to.
 
> Docker is how we get a hermetic, repeatable build.

I agree with these two statements, and they're not mutually exclusive.

> That should be the default.

If it's easy to do it either way, I don't really care about this.

> Everyone should be using containers for build.

I strongly disagree with this. If I want to just use `go build` and `go test` for local development, I should be able to. If I want to forgo the pains of docker, let me. Please do not actively prevent people from doing things like this. Please make it easy to actually do this.


My opinion: local builds should be quick, easy, and pretty much like building Go anywhere else. If you want hermetic local builds, that should also be easy. CI builders should of course be hermetic.

**Comment by thockin** [comment]
At this point "go build" and "go test" are laughably underpowered.  I don't
think we should actively prevent local builds, but we should not endorse it
as best practice, and we should not go very far out of our way to make it
better.

**Comment by krousey** [comment]
> At this point "go build" and "go test" are laughably underpowered.

Could you please elaborate? Because honestly this is 85%+ of my daily workflow. Outside of packaging for `kube-up.sh` or e2e tests, I don't really use our `make` infrastructure.

Also, I'm not asking for it to be endorsed as the official way to build. I am asking for it to be supported though. I think it would be a great disservice to break the basic tools that many developers use in different ways. For example, breaking `go build` and `go test` for me would probably break: IDE on-the-fly building and testing, AST-aware autocompletion, AST-aware code refactoring, AST-aware code navigation, and probably a lot of other tools that I don't even know I use day to day.

**Comment by thockin** [comment]
On Wed, Aug 24, 2016 at 2:50 PM, krousey <notifications@github.com> wrote:
>
> At this point "go build" and "go test" are laughably underpowered.
>
> Could you please elaborate? Because honestly this is 85%+ of my daily workflow. Outside of packaging for kube-up.sh or e2e tests, I don't really use our make infrastructure.

We already have code files that are generated automatically on-demand.
If you don't regenerate them when you need to, you might have false
positives or negatives.  If you're OK with being The Human Makefile,
`go build` might work.  For now.  Right up until we stop checking the
generated files in, at which case you might fail to compile or run.

TL;DR: the supported way to build is "make".

> Also, I'm not asking for it to be endorsed as the official way to build. I am asking for it to be supported though. I think it would be a great disservice to break the basic tools that many developers use in different ways. For example, breaking go build and go test for me would probably break: IDE on-the-fly building and testing, AST-aware autocompletion, AST-aware code refactoring, AST-aware code navigation, and probably a lot of other tools that I don't even know I use day to day.

It's already broken, just subtly so.  I'm all ear for better answers,
but I know you are first-hand familiar with the rebase hell that we
have today.

**Comment by spxtr** [comment]
I'm fine with using `make`, but I would like a make rule (`make local` or something) that does the entire build locally and respects my environment. It should make sure that the generated code is generated at the right times before calling `go install`. If this is the case then go tools will still work properly as long as you call `make generated_files` when necessary (The Human Makefile!), and you'll get fast incremental builds. Developers should use `make local` for their work.

Then we can have another rule that does the above build within a container. The only code that should be executing outside of the container there is the code required to deal with docker. Maybe we can do some work to make this use a data container and be fast for iterative development, but that might be more work than it's worth.

I think this is the best compromise and it's conceptually simple for everyone. I had a working example in the PR I linked above.

**Comment by thockin** [comment]
As this is all hypothetical still, easy to agree on.

On Tue, Aug 30, 2016 at 1:39 PM, Joe Finney <notifications@github.com> wrote:
>
> I'm fine with using make, but I would like a make rule (make local or something) that does the entire build locally and respects my environment. It should make sure that the generated code is generated at the right times before calling go install.

I'm fine with that

> If this is the case then go tools will still work properly as long as you call make generated_files when necessary (The Human Makefile!), and you'll get fast incremental builds.

Humans should never NEED to call anything but 'make' or 'make LOCAL=1'
or similar.

> Developers should use make local for their work.

?!  Disagree.  I honestly see NO REASON not to use docker for builds.
Local support is for crazies who think they can have workstations
without docker installed, and are willing to install ALL the tools we
need for install.

**Comment by spxtr** [comment]
> Humans should never NEED to call anything but 'make' or 'make LOCAL=1'
> or similar.

Agreed.

> ?!  Disagree.  I honestly see NO REASON not to use docker for builds.
> Local support is for crazies who think they can have workstations
> without docker installed, and are willing to install ALL the tools we
> need for install.

If you're compiling for linux/amd64 then there really aren't any unexpected dependencies for building. It's pretty much just Go, Docker, and standard build essentials. The main reason not to use docker for builds is that it's confusing and complicated, and it's going to be buggy and difficult to fix when it breaks. #30787 demonstrates that getting a fast, iterative, dockerized build is super tricky.

Interestingly, one way to get fast dockerized builds is to mount the entire GOPATH into the container instead of just the kubernetes repo. If the user hasn't set up their GOPATH properly then just mount the repo and you have a non-iterative build.

**Comment by thockin** [comment]
To build kube you need tools like gogoproto, at least.  That might be
vendored - we have not been exactly consistent about that.

WRT You can mount, relatively easily, enough information to have a fully
incremental, multi-arch-capable build system with docker.  You just have to
abandon MacOS users (or punish them for their bad decisions).  That whole
mess with rsyc and data volumes is a complete hacktastic disaster, IMO.

**Comment by ixdy** [comment]
Please add things I'm missing to https://github.com/kubernetes/kubernetes/projects/3!

**Comment by spxtr** [comment]
Okay, I'm mostly convinced that bazel is the correct approach here. It has some awkward points but I think the benefits are major enough to warrant it.

Downsides:

1. Lots of auto-generated BUILD rules, some tricky edge cases involved with that. Gazelle isn't quite ready yet so we have to roll our own answer until it is.
1. It's unfamiliar to non-googlers and isn't trivial to install on certain systems.

Upsides:

1. It's an actual build system and not a kludge of make and bash. Importantly, it is a separately-maintained project. We don't need to waste so much of our time on it.
1. Can be run in docker, on kubernetes, whatever. It sandboxes builds by default.
1. Can generate docker images and tar archives along with go packages and binaries.
1. Caches test binaries and results along with regular binaries and packages. This is super nice, since building our tests takes a very long time. We have `cmd/teststale` to help, but it's not enough.
1. Produces JUnit XML output. No more go-junit-report.

**Comment by thockin** [comment]
I wish to resign from the build business.  I find it ugly but if it works
and doesn't require me to debug the intersection of `go install` and make
DAGs, I'll get over the ugliness.

**Comment by jbeda** [comment]
I'm not excited about bazel but, like @thockin, I'm not excited about making builds my life.  A combination of dockerized builds and bazel might be a good combo.  (I say this as using a dockerized build might help to mitigate the pain of installing bazel.)

@spxtr -- do you have the bandwidth to try and convert things to bazel to see how it works?

**Comment by krousey** [comment]
@jbeda #32774 is @mikedanese 's stab at it.

**Comment by mikedanese** [comment]
It works well :) just need buy in and polish.

**Comment by ixdy** [comment]
We're planning to move forward with Bazel. As @spxtr notes there are quite a few advantages which should simplify a lot of the maintenance and replace the hacks we currently have in our build/release infrastructure.

Rough plan:
* merge #32774, along with verify or autogen scripts necessary to keep `bazel build` working. The officially supported way to build and release (using docker and the `build` scripts.
) will remain as-is until at least 1.6.
* set up separate non-blocking CI jobs to run `bazel build` and `bazel test`
* TBD: generated files?
* create a docker image and/or scripts to allow folks to run `bazel` inside docker (since it's less easy to install natively unless you have Ubuntu or OSX)
* set up bazel to produce all of the kubernetes release artifacts (tarballs, docker images, etc) for one arch (`linux/amd64`). probably make this optional for 1.5. 
* get cross-compilation working in bazel, and start producing release artifacts for all architectures
* write `make` wrappers to `bazel` commands to keep @thockin happy
* ???
* profit (delete `build/` scripts) - 1.6 at the earliest

**Comment by ixdy** [comment]
A few other things we'll need to figure out (or wait for bazel to support):
* go build tags
* race detection
* code coverage

**Comment by spxtr** [comment]
I'd be happy to get the Jenkins stuff set up.

**Comment by ixdy** [comment]
Implied but missing from my list: community feedback and acceptance. Probably step 2 or 3.

**Comment by smarterclayton** [comment]
Comment from the other issue is that it's unclear to me how bazel can fit within the distribution ecosystem (fedora/centos/rhel/debian/ubuntu/others) and what sorts of changes are implied or whether it runs afoul of any packaging guidelines that are going to complicate this.  Without that step, kubernetes developers might be more productive, but everyone else is more broken.

@eparis / @ingvagabund / @dgoodwin can you guys get familiar enough with bazel to weigh in on the impacts to koji/centos build systems?

**Comment by ixdy** [comment]
Can someone point me at the scripts used by various distros to build kubernetes from source?

**Comment by ingvagabund** [comment]
Fedora: http://pkgs.fedoraproject.org/cgit/rpms/kubernetes.git/tree/kubernetes.spec?id=6fad2e50ff2a4dca3655176ad935012dd271e775#n853

**Comment by spxtr** [comment]
I think we can close this. Please open new issues for any issues with Bazel.
